File: CHUONG_7/7.09_MultiThread/app/src/main/java/com/example/myapplication/viewmodel/MainActivityViewModel.java
package com.example.myapplication.viewmodel;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.myapplication.model.Song;
import com.example.myapplication.model.SongList;
import com.example.myapplication.repository.LoadSongRepository;
import com.example.myapplication.resource.Result;

import java.util.List;

public class MainActivityViewModel extends ViewModel {
    private final LoadSongRepository repository;
    private final MutableLiveData<List<Song>> _songs;
    public LiveData<List<Song>> songs;

    public MainActivityViewModel(LoadSongRepository repository){
        this.repository = repository;
        this._songs = new MutableLiveData<>();
        this.songs = _songs;
        loadSongs();
    }

    public void loadSongs() {
        repository.makeGetSongRequest(result -> {
            if (result instanceof Result.Success){
                _songs.postValue(((Result.Success<SongList>) result).data.songs);
            } else {
                _songs.postValue(null);
            }
        });
    }
}

File: CHUONG_7/7.08_Coroutine/app/src/main/java/com/example/coroutine/viewmodel/MainActivityViewModel.kt
package com.example.coroutine.viewmodel

import androidx.lifecycle.LiveData
import com.example.coroutine.resource.Result
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.example.coroutine.model.Song
import com.example.coroutine.repository.LoadSongRepository
import com.example.coroutine.resource.SongResponseParser
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.launch

class MainActivityViewModel(
    private val parser: SongResponseParser = SongResponseParser(),
    private val repository: LoadSongRepository = LoadSongRepository(parser)
) : ViewModel() {
    private val _songList = MutableLiveData<List<Song>?>()
    val songList: LiveData<List<Song>?>
        get() = _songList

    init {
        load()
    }

    fun load() {
        viewModelScope.launch {
            val result = repository.loadSongs()
            if (result is Result.Success) {
                _songList.postValue(result.data.music)
            } else {
                _songList.postValue(null)
            }
        }
    }
}